anvil/observables.py

```diff
@@ -153,16 +148,24 @@ def bootstrap_function(func, states, *args, n_boot=100):
         dimension
 
     """
-    boot_index = torch.randint(0, states.shape[0], size=(states.shape[0], n_boot))
-    # put boot index on final dimension
-    resampled_states = states[boot_index, :].transpose(1, 2)
-    res = func(resampled_states, *args)
-    return res
+    # NOTE: Optimal seems to be around n_sample x batch_size = 10000 for
+    # calc_two_point_function on my laptop. Clearly this is not a general
+    # rule though.
+    sample_size = states.shape[0]
+    batch_size = min(n_boot, max(1, 10000 // sample_size))
+    n_batch = n_boot // batch_size
 
+    pbar = tqdm(range(n_batch), desc="bootstrap batch")
```

Comment:
I kind of hate using tqdm in sampling/observables but I guess it's useful when something takes minutes that you were expecting to take seconds, whats the longest you would expect to be in this function now?

Reply:
just wonder if we could `log.debug(f"bootstrap batch: {n}")` in loop so when its working fine we dont need to see this and then can run in debug mode when it takes ages to see whats happening?

Reply:
> I kind of hate using tqdm in sampling/observables

I see the little white bar and I feel all is right with the world.

> whats the longest you would expect to be in this function now?

Hmm it's actually not scaling too well at the moment. For target length 10000 and 1000 bootstrap samples I get 50s, 2 mins, 5 mins for L = 6, 8, 10. I will look into this.

> just wonder if we could `log.debug(f"bootstrap batch: {n}")` in loop so when its working fine we dont need to see this and then can run in debug mode when it takes ages to see whats happening?

I'm fine with that.

Reply:
> Hmm it's actually not scaling too well at the moment. For target length 10000 and 1000 bootstrap samples I get 50s, 2 mins, 5 mins for L = 6, 8, 10. I will look into this.

I'm being dumb. This is just because `calc_two_point_function` has much more work to do.